In [1]:
import pandas as pd
import numpy as np

In [2]:
stat = pd.read_csv('https://gbcdn.mrgcdn.ru/uploads/asset/4248710/attachment/5f6b595e63350aa5b55f0ea72561b871.csv',sep=';')

In [3]:
stat.head()

Company      Product   TypeName  Inches                    ScreenResolution  \
0   Apple  MacBook Pro  Ultrabook    13.3  IPS Panel Retina Display 2560x1600   
1   Apple  Macbook Air  Ultrabook    13.3                            1440x900   
2      HP       250 G6   Notebook    15.6                   Full HD 1920x1080   
3   Apple  MacBook Pro  Ultrabook    15.4  IPS Panel Retina Display 2880x1800   
4   Apple  MacBook Pro  Ultrabook    13.3  IPS Panel Retina Display 2560x1600   

                          Cpu   Ram               Memory  \
0        Intel Core i5 2.3GHz   8GB            128GB SSD   
1        Intel Core i5 1.8GHz   8GB  128GB Flash Storage   
2  Intel Core i5 7200U 2.5GHz   8GB            256GB SSD   
3        Intel Core i7 2.7GHz  16GB            512GB SSD   
4        Intel Core i5 3.1GHz   8GB            256GB SSD   

                            Gpu  OpSys  Weight  Price_euros  
0  Intel Iris Plus Graphics 640  macOS  1.37kg      1339.69  
1        Intel HD Graphics 6000  macOS  1.34kg       898.94  
2         Intel HD Graphics 620  No OS  1.86kg       575.00  
3            AMD Radeon Pro 455  macOS  1.83kg      2537.45  
4  Intel Iris Plus Graphics 650  macOS  1.37kg      1803.60

## 1.1 Создать новый признак Cpu_Company, который будет содержать только название фирмы, которая произвела CPU

In [4]:
stat['Cpu_Company'] = 'Company'
stat.head()

Company      Product   TypeName  Inches                    ScreenResolution  \
0   Apple  MacBook Pro  Ultrabook    13.3  IPS Panel Retina Display 2560x1600   
1   Apple  Macbook Air  Ultrabook    13.3                            1440x900   
2      HP       250 G6   Notebook    15.6                   Full HD 1920x1080   
3   Apple  MacBook Pro  Ultrabook    15.4  IPS Panel Retina Display 2880x1800   
4   Apple  MacBook Pro  Ultrabook    13.3  IPS Panel Retina Display 2560x1600   

                          Cpu   Ram               Memory  \
0        Intel Core i5 2.3GHz   8GB            128GB SSD   
1        Intel Core i5 1.8GHz   8GB  128GB Flash Storage   
2  Intel Core i5 7200U 2.5GHz   8GB            256GB SSD   
3        Intel Core i7 2.7GHz  16GB            512GB SSD   
4        Intel Core i5 3.1GHz   8GB            256GB SSD   

                            Gpu  OpSys  Weight  Price_euros Cpu_Company  
0  Intel Iris Plus Graphics 640  macOS  1.37kg      1339.69     Company  
1        Intel HD Graphics 6000  macOS  1.34kg       898.94     Company  
2         Intel HD Graphics 620  No OS  1.86kg       575.00     Company  
3            AMD Radeon Pro 455  macOS  1.83kg      2537.45     Company  
4  Intel Iris Plus Graphics 650  macOS  1.37kg      1803.60     Company

In [5]:
stat['Cpu_Company'] = stat['Cpu']
stat.head()

Company      Product   TypeName  Inches                    ScreenResolution  \
0   Apple  MacBook Pro  Ultrabook    13.3  IPS Panel Retina Display 2560x1600   
1   Apple  Macbook Air  Ultrabook    13.3                            1440x900   
2      HP       250 G6   Notebook    15.6                   Full HD 1920x1080   
3   Apple  MacBook Pro  Ultrabook    15.4  IPS Panel Retina Display 2880x1800   
4   Apple  MacBook Pro  Ultrabook    13.3  IPS Panel Retina Display 2560x1600   

                          Cpu   Ram               Memory  \
0        Intel Core i5 2.3GHz   8GB            128GB SSD   
1        Intel Core i5 1.8GHz   8GB  128GB Flash Storage   
2  Intel Core i5 7200U 2.5GHz   8GB            256GB SSD   
3        Intel Core i7 2.7GHz  16GB            512GB SSD   
4        Intel Core i5 3.1GHz   8GB            256GB SSD   

                            Gpu  OpSys  Weight  Price_euros  \
0  Intel Iris Plus Graphics 640  macOS  1.37kg      1339.69   
1        Intel HD Graphics 6000  macOS  1.34kg       898.94   
2         Intel HD Graphics 620  No OS  1.86kg       575.00   
3            AMD Radeon Pro 455  macOS  1.83kg      2537.45   
4  Intel Iris Plus Graphics 650  macOS  1.37kg      1803.60   

                  Cpu_Company  
0        Intel Core i5 2.3GHz  
1        Intel Core i5 1.8GHz  
2  Intel Core i5 7200U 2.5GHz  
3        Intel Core i7 2.7GHz  
4        Intel Core i5 3.1GHz

In [6]:
def separator (value):
    new_list = []
    for i in value:
        while  i != ' ':
            new_list.append(i)
            break 
        if i == ' ':
            break
    result = ""
    for g in new_list:
        result = result + g
    return result

In [7]:
stat['Cpu_Company'] = stat['Cpu_Company'].apply(separator)

In [8]:
stat.head()

Company      Product   TypeName  Inches                    ScreenResolution  \
0   Apple  MacBook Pro  Ultrabook    13.3  IPS Panel Retina Display 2560x1600   
1   Apple  Macbook Air  Ultrabook    13.3                            1440x900   
2      HP       250 G6   Notebook    15.6                   Full HD 1920x1080   
3   Apple  MacBook Pro  Ultrabook    15.4  IPS Panel Retina Display 2880x1800   
4   Apple  MacBook Pro  Ultrabook    13.3  IPS Panel Retina Display 2560x1600   

                          Cpu   Ram               Memory  \
0        Intel Core i5 2.3GHz   8GB            128GB SSD   
1        Intel Core i5 1.8GHz   8GB  128GB Flash Storage   
2  Intel Core i5 7200U 2.5GHz   8GB            256GB SSD   
3        Intel Core i7 2.7GHz  16GB            512GB SSD   
4        Intel Core i5 3.1GHz   8GB            256GB SSD   

                            Gpu  OpSys  Weight  Price_euros Cpu_Company  
0  Intel Iris Plus Graphics 640  macOS  1.37kg      1339.69       Intel  
1        Intel HD Graphics 6000  macOS  1.34kg       898.94       Intel  
2         Intel HD Graphics 620  No OS  1.86kg       575.00       Intel  
3            AMD Radeon Pro 455  macOS  1.83kg      2537.45       Intel  
4  Intel Iris Plus Graphics 650  macOS  1.37kg      1803.60       Intel

## 1.2 Создать новый признак Memory_Amount, который будет содержать только количество Gb памяти без указания типа носителя

In [9]:
stat['Memory_Amount'] = 0

In [10]:
## stat['Memory_Amount'] = stat['Memory'].apply(separator)

In [11]:
def convert_to_gb(x):
    memory = x.split(' ')[0]
    if memory.endswith('GB'):
        res = memory.replace('GB', '')
    elif memory.endswith('TB'):
        res = float(memory.replace ('TB', ''))*1024
    return int(res)

In [12]:
stat['Memory_Amount'] = stat['Memory'].apply(convert_to_gb)

In [13]:
stat.head()

Company      Product   TypeName  Inches                    ScreenResolution  \
0   Apple  MacBook Pro  Ultrabook    13.3  IPS Panel Retina Display 2560x1600   
1   Apple  Macbook Air  Ultrabook    13.3                            1440x900   
2      HP       250 G6   Notebook    15.6                   Full HD 1920x1080   
3   Apple  MacBook Pro  Ultrabook    15.4  IPS Panel Retina Display 2880x1800   
4   Apple  MacBook Pro  Ultrabook    13.3  IPS Panel Retina Display 2560x1600   

                          Cpu   Ram               Memory  \
0        Intel Core i5 2.3GHz   8GB            128GB SSD   
1        Intel Core i5 1.8GHz   8GB  128GB Flash Storage   
2  Intel Core i5 7200U 2.5GHz   8GB            256GB SSD   
3        Intel Core i7 2.7GHz  16GB            512GB SSD   
4        Intel Core i5 3.1GHz   8GB            256GB SSD   

                            Gpu  OpSys  Weight  Price_euros Cpu_Company  \
0  Intel Iris Plus Graphics 640  macOS  1.37kg      1339.69       Intel   
1        Intel HD Graphics 6000  macOS  1.34kg       898.94       Intel   
2         Intel HD Graphics 620  No OS  1.86kg       575.00       Intel   
3            AMD Radeon Pro 455  macOS  1.83kg      2537.45       Intel   
4  Intel Iris Plus Graphics 650  macOS  1.37kg      1803.60       Intel   

   Memory_Amount  
0            128  
1            128  
2            256  
3            512  
4            256

## 1.3 Создать новый признак Memory_Type, который будет содержать только тип носителя (HDD/SDD/др.)

In [14]:
stat['Memory_Type'] = 0
stat.head()

Company      Product   TypeName  Inches                    ScreenResolution  \
0   Apple  MacBook Pro  Ultrabook    13.3  IPS Panel Retina Display 2560x1600   
1   Apple  Macbook Air  Ultrabook    13.3                            1440x900   
2      HP       250 G6   Notebook    15.6                   Full HD 1920x1080   
3   Apple  MacBook Pro  Ultrabook    15.4  IPS Panel Retina Display 2880x1800   
4   Apple  MacBook Pro  Ultrabook    13.3  IPS Panel Retina Display 2560x1600   

                          Cpu   Ram               Memory  \
0        Intel Core i5 2.3GHz   8GB            128GB SSD   
1        Intel Core i5 1.8GHz   8GB  128GB Flash Storage   
2  Intel Core i5 7200U 2.5GHz   8GB            256GB SSD   
3        Intel Core i7 2.7GHz  16GB            512GB SSD   
4        Intel Core i5 3.1GHz   8GB            256GB SSD   

                            Gpu  OpSys  Weight  Price_euros Cpu_Company  \
0  Intel Iris Plus Graphics 640  macOS  1.37kg      1339.69       Intel   
1        Intel HD Graphics 6000  macOS  1.34kg       898.94       Intel   
2         Intel HD Graphics 620  No OS  1.86kg       575.00       Intel   
3            AMD Radeon Pro 455  macOS  1.83kg      2537.45       Intel   
4  Intel Iris Plus Graphics 650  macOS  1.37kg      1803.60       Intel   

   Memory_Amount  Memory_Type  
0            128            0  
1            128            0  
2            256            0  
3            512            0  
4            256            0

In [15]:
# def separator_2 (value):
#     new_list = []
#     for i in value:
#         while  i != ' ':
#             continue
#         else:
#             new_list.append(i)
#         break
#         if i == ' ':
#             break
#     result = ""
#     for g in new_list:
#         result = result + g
#     return result

In [16]:
def convert_to_type(x):
    memory_type = x.replace(' ','').split('B')
    return memory_type[1]

In [17]:
stat['Memory_Type'] = stat['Memory'].apply(convert_to_type)

In [18]:
stat.head()

Company      Product   TypeName  Inches                    ScreenResolution  \
0   Apple  MacBook Pro  Ultrabook    13.3  IPS Panel Retina Display 2560x1600   
1   Apple  Macbook Air  Ultrabook    13.3                            1440x900   
2      HP       250 G6   Notebook    15.6                   Full HD 1920x1080   
3   Apple  MacBook Pro  Ultrabook    15.4  IPS Panel Retina Display 2880x1800   
4   Apple  MacBook Pro  Ultrabook    13.3  IPS Panel Retina Display 2560x1600   

                          Cpu   Ram               Memory  \
0        Intel Core i5 2.3GHz   8GB            128GB SSD   
1        Intel Core i5 1.8GHz   8GB  128GB Flash Storage   
2  Intel Core i5 7200U 2.5GHz   8GB            256GB SSD   
3        Intel Core i7 2.7GHz  16GB            512GB SSD   
4        Intel Core i5 3.1GHz   8GB            256GB SSD   

                            Gpu  OpSys  Weight  Price_euros Cpu_Company  \
0  Intel Iris Plus Graphics 640  macOS  1.37kg      1339.69       Intel   
1        Intel HD Graphics 6000  macOS  1.34kg       898.94       Intel   
2         Intel HD Graphics 620  No OS  1.86kg       575.00       Intel   
3            AMD Radeon Pro 455  macOS  1.83kg      2537.45       Intel   
4  Intel Iris Plus Graphics 650  macOS  1.37kg      1803.60       Intel   

   Memory_Amount   Memory_Type  
0            128           SSD  
1            128  FlashStorage  
2            256           SSD  
3            512           SSD  
4            256           SSD

## 1.4 Удалите признаки Memory и ScreenResolution

In [19]:
stat.drop(columns = 'Memory', inplace= True)

In [20]:
stat.drop(columns = 'ScreenResolution', inplace= True)

In [21]:
stat.head()

Company      Product   TypeName  Inches                         Cpu   Ram  \
0   Apple  MacBook Pro  Ultrabook    13.3        Intel Core i5 2.3GHz   8GB   
1   Apple  Macbook Air  Ultrabook    13.3        Intel Core i5 1.8GHz   8GB   
2      HP       250 G6   Notebook    15.6  Intel Core i5 7200U 2.5GHz   8GB   
3   Apple  MacBook Pro  Ultrabook    15.4        Intel Core i7 2.7GHz  16GB   
4   Apple  MacBook Pro  Ultrabook    13.3        Intel Core i5 3.1GHz   8GB   

                            Gpu  OpSys  Weight  Price_euros Cpu_Company  \
0  Intel Iris Plus Graphics 640  macOS  1.37kg      1339.69       Intel   
1        Intel HD Graphics 6000  macOS  1.34kg       898.94       Intel   
2         Intel HD Graphics 620  No OS  1.86kg       575.00       Intel   
3            AMD Radeon Pro 455  macOS  1.83kg      2537.45       Intel   
4  Intel Iris Plus Graphics 650  macOS  1.37kg      1803.60       Intel   

   Memory_Amount   Memory_Type  
0            128           SSD  
1            128  FlashStorage  
2            256           SSD  
3            512           SSD  
4            256           SSD

# Задание 2
## 2.1 Создайте признак SSD, который изначально равен 0

In [22]:
stat['SSD'] = 0

In [23]:
stat.head()

Company      Product   TypeName  Inches                         Cpu   Ram  \
0   Apple  MacBook Pro  Ultrabook    13.3        Intel Core i5 2.3GHz   8GB   
1   Apple  Macbook Air  Ultrabook    13.3        Intel Core i5 1.8GHz   8GB   
2      HP       250 G6   Notebook    15.6  Intel Core i5 7200U 2.5GHz   8GB   
3   Apple  MacBook Pro  Ultrabook    15.4        Intel Core i7 2.7GHz  16GB   
4   Apple  MacBook Pro  Ultrabook    13.3        Intel Core i5 3.1GHz   8GB   

                            Gpu  OpSys  Weight  Price_euros Cpu_Company  \
0  Intel Iris Plus Graphics 640  macOS  1.37kg      1339.69       Intel   
1        Intel HD Graphics 6000  macOS  1.34kg       898.94       Intel   
2         Intel HD Graphics 620  No OS  1.86kg       575.00       Intel   
3            AMD Radeon Pro 455  macOS  1.83kg      2537.45       Intel   
4  Intel Iris Plus Graphics 650  macOS  1.37kg      1803.60       Intel   

   Memory_Amount   Memory_Type  SSD  
0            128           SSD    0  
1            128  FlashStorage    0  
2            256           SSD    0  
3            512           SSD    0  
4            256           SSD    0

## 2.2 Поставьте в признаке SSD 1, если ноутбук действительно с типом носителя SSD

In [24]:
stat.loc[stat['Memory_Type'] == 'SSD','SSD'] = 1

In [25]:
stat.tail()

Company                              Product            TypeName  Inches  \
1298  Lenovo                       Yoga 500-14ISK  2 in 1 Convertible    14.0   
1299  Lenovo                       Yoga 900-13ISK  2 in 1 Convertible    13.3   
1300  Lenovo                   IdeaPad 100S-14IBR            Notebook    14.0   
1301      HP  15-AC110nv (i7-6500U/6GB/1TB/Radeon            Notebook    15.6   
1302    Asus  X553SA-XX031T (N3050/4GB/500GB/W10)            Notebook    15.6   

                                       Cpu   Ram                    Gpu  \
1298            Intel Core i7 6500U 2.5GHz   4GB  Intel HD Graphics 520   
1299            Intel Core i7 6500U 2.5GHz  16GB  Intel HD Graphics 520   
1300  Intel Celeron Dual Core N3050 1.6GHz   2GB      Intel HD Graphics   
1301            Intel Core i7 6500U 2.5GHz   6GB     AMD Radeon R5 M330   
1302  Intel Celeron Dual Core N3050 1.6GHz   4GB      Intel HD Graphics   

           OpSys  Weight  Price_euros Cpu_Company  Memory_Amount  \
1298  Windows 10   1.8kg        638.0       Intel            128   
1299  Windows 10   1.3kg       1499.0       Intel            512   
1300  Windows 10   1.5kg        229.0       Intel             64   
1301  Windows 10  2.19kg        764.0       Intel           1024   
1302  Windows 10   2.2kg        369.0       Intel            500   

       Memory_Type  SSD  
1298           SSD    1  
1299           SSD    1  
1300  FlashStorage    0  
1301           HDD    0  
1302           HDD    0

## 2.3 Уберите в признаке Weight значения 'kg' и поменяйте его тип данных на вещественный

In [26]:
def weight_replace (x):
    weight = x.replace('kg','')
    return weight

In [27]:
stat['Weight'] = stat['Weight'].apply(weight_replace)

In [28]:
stat.head()

Company      Product   TypeName  Inches                         Cpu   Ram  \
0   Apple  MacBook Pro  Ultrabook    13.3        Intel Core i5 2.3GHz   8GB   
1   Apple  Macbook Air  Ultrabook    13.3        Intel Core i5 1.8GHz   8GB   
2      HP       250 G6   Notebook    15.6  Intel Core i5 7200U 2.5GHz   8GB   
3   Apple  MacBook Pro  Ultrabook    15.4        Intel Core i7 2.7GHz  16GB   
4   Apple  MacBook Pro  Ultrabook    13.3        Intel Core i5 3.1GHz   8GB   

                            Gpu  OpSys Weight  Price_euros Cpu_Company  \
0  Intel Iris Plus Graphics 640  macOS   1.37      1339.69       Intel   
1        Intel HD Graphics 6000  macOS   1.34       898.94       Intel   
2         Intel HD Graphics 620  No OS   1.86       575.00       Intel   
3            AMD Radeon Pro 455  macOS   1.83      2537.45       Intel   
4  Intel Iris Plus Graphics 650  macOS   1.37      1803.60       Intel   

   Memory_Amount   Memory_Type  SSD  
0            128           SSD    1  
1            128  FlashStorage    0  
2            256           SSD    1  
3            512           SSD    1  
4            256           SSD    1

In [29]:
stat['Weight'] = stat['Weight'].astype('float')

In [30]:
print(type(stat['Weight'][0]))

<class 'numpy.float64'>


# Задание 3
## Создайте DataFrame

In [31]:
clients = pd.DataFrame({
    'client_id': [45, 32, 67, 33, 43],
    'laptop_id': [506, 398, 710, 120, 1999]
})

In [32]:
clients.head()

client_id  laptop_id
0         45        506
1         32        398
2         67        710
3         33        120
4         43       1999

## 3.1 Присоедините к таблице clients данные по ноутбукам через метод join

In [33]:
clients_id = clients.set_index('laptop_id')

In [34]:
joined = clients.join(stat)

In [35]:
joined

client_id  laptop_id Company      Product   TypeName  Inches  \
0         45        506   Apple  MacBook Pro  Ultrabook    13.3   
1         32        398   Apple  Macbook Air  Ultrabook    13.3   
2         67        710      HP       250 G6   Notebook    15.6   
3         33        120   Apple  MacBook Pro  Ultrabook    15.4   
4         43       1999   Apple  MacBook Pro  Ultrabook    13.3   

                          Cpu   Ram                           Gpu  OpSys  \
0        Intel Core i5 2.3GHz   8GB  Intel Iris Plus Graphics 640  macOS   
1        Intel Core i5 1.8GHz   8GB        Intel HD Graphics 6000  macOS   
2  Intel Core i5 7200U 2.5GHz   8GB         Intel HD Graphics 620  No OS   
3        Intel Core i7 2.7GHz  16GB            AMD Radeon Pro 455  macOS   
4        Intel Core i5 3.1GHz   8GB  Intel Iris Plus Graphics 650  macOS   

   Weight  Price_euros Cpu_Company  Memory_Amount   Memory_Type  SSD  
0    1.37      1339.69       Intel            128           SSD    1  
1    1.34       898.94       Intel            128  FlashStorage    0  
2    1.86       575.00       Intel            256           SSD    1  
3    1.83      2537.45       Intel            512           SSD    1  
4    1.37      1803.60       Intel            256           SSD    1

## 3.2 Присоедините к таблице clients данные по ноутбукам через метод merge

In [36]:
clients.head()

client_id  laptop_id
0         45        506
1         32        398
2         67        710
3         33        120
4         43       1999

In [37]:
stat['laptop_id'] = stat.index

In [38]:
merged = clients.merge(stat, on  = 'laptop_id')
merged.head()

client_id  laptop_id Company                 Product     TypeName  Inches  \
0         45        506    Asus  ZenBook UX510UX-CN211T     Notebook    15.6   
1         32        398    Dell         Precision M5520  Workstation    15.6   
2         67        710  Lenovo      Legion Y520-15IKBN       Gaming    15.6   
3         33        120    Acer                  Spin 3     Notebook    15.6   

                           Cpu  Ram                         Gpu       OpSys  \
0   Intel Core i7 7500U 2.7GHz  8GB       Intel HD Graphics 620  Windows 10   
1  Intel Core i7 7700HQ 2.8GHz  8GB         Nvidia Quadro M1200  Windows 10   
2  Intel Core i7 7700HQ 2.8GHz  8GB  Nvidia GeForce GTX 1050 Ti  Windows 10   
3   Intel Core i3 7100U 2.4GHz  6GB       Intel HD Graphics 620  Windows 10   

   Weight  Price_euros Cpu_Company  Memory_Amount Memory_Type  SSD  
0    2.00       1224.0       Intel            256      SSD+1T    0  
1    1.78       2712.0       Intel            256         SSD    1  
2    2.50       1249.0       Intel            128      SSD+1T    0  
3    2.10        479.0       Intel           1024         HDD    0

# Задание 4
## 4.1 Найдите среднюю стоимость ноутбуков в зависимости от компании производителя

In [39]:
stat.head()

Company      Product   TypeName  Inches                         Cpu   Ram  \
0   Apple  MacBook Pro  Ultrabook    13.3        Intel Core i5 2.3GHz   8GB   
1   Apple  Macbook Air  Ultrabook    13.3        Intel Core i5 1.8GHz   8GB   
2      HP       250 G6   Notebook    15.6  Intel Core i5 7200U 2.5GHz   8GB   
3   Apple  MacBook Pro  Ultrabook    15.4        Intel Core i7 2.7GHz  16GB   
4   Apple  MacBook Pro  Ultrabook    13.3        Intel Core i5 3.1GHz   8GB   

                            Gpu  OpSys  Weight  Price_euros Cpu_Company  \
0  Intel Iris Plus Graphics 640  macOS    1.37      1339.69       Intel   
1        Intel HD Graphics 6000  macOS    1.34       898.94       Intel   
2         Intel HD Graphics 620  No OS    1.86       575.00       Intel   
3            AMD Radeon Pro 455  macOS    1.83      2537.45       Intel   
4  Intel Iris Plus Graphics 650  macOS    1.37      1803.60       Intel   

   Memory_Amount   Memory_Type  SSD  laptop_id  
0            128           SSD    1          0  
1            128  FlashStorage    0          1  
2            256           SSD    1          2  
3            512           SSD    1          3  
4            256           SSD    1          4

In [40]:
stat.groupby('Company').agg({'Price_euros': 'mean'}).sort_values('Price_euros')

Price_euros
Company               
Vero        217.425000
Mediacom    295.000000
Chuwi       314.296667
Acer        626.775825
Fujitsu     729.000000
HP         1067.774854
Lenovo     1086.384444
Asus       1104.169367
Xiaomi     1133.462500
Dell       1186.068990
Toshiba    1267.812500
Samsung    1413.444444
Huawei     1424.000000
Apple      1564.198571
Microsoft  1612.308333
Google     1677.666667
MSI        1728.908148
LG         2099.000000
Razer      3346.142857

## 4.2 Найдите минимальную, среднюю и максимальную стоимости ноутбуков в зависимости от производителя процессора

In [41]:
stat.groupby('Cpu_Company').agg({'Price_euros': ['min', 'mean', 'max']})

Price_euros                     
                    min         mean     max
Cpu_Company                                 
AMD               199.0   560.638871  2199.0
Intel             174.0  1152.214145  6099.0
Samsung           659.0   659.000000   659.0

## 4.3 Постройте таблицу с подсчетом количества ноутбуков в данных в зависимости от производителя CPU и ОЗУ

In [42]:
stat.pivot_table(index = 'Cpu_Company',columns = 'Ram',
                values = 'Memory_Type', aggfunc = 'count',
                 fill_value=0)

Ram          12GB  16GB  24GB  2GB  32GB  4GB  64GB  6GB  8GB
Cpu_Company                                                  
AMD             2     3     0    1     0   31     0   13   12
Intel          23   197     3   21    17  343     1   28  607
Samsung         0     0     0    0     0    1     0    0    0

## 4.4 Постройте таблицу с подсчетом средней стоимости ноутбуков в данных в зависимости от операционной системы и GB памяти

In [43]:
stat.pivot_table(index = 'OpSys', columns = 'Memory_Amount',
                values = 'Price_euros', aggfunc = 'mean',
                fill_value = '-')

Memory_Amount    8       16          32          64           128     180   \
OpSys                                                                        
Android             -       -           -       434.0            -       -   
Chrome OS           -  305.38  412.454545  774.333333       1275.0       -   
Linux               -       -           -           -       742.25       -   
Mac OS X            -       -           -           -       1099.0       -   
No OS               -       -           -           -       562.14       -   
Windows 10     2249.0       -  270.001471     499.716  1026.489167  1073.5   
Windows 10 S        -       -           -     308.995       1039.0       -   
Windows 7           -       -           -           -  1320.323333  1199.0   
macOS               -       -           -           -     1119.315       -   

Memory_Amount    240          256         500     508          512   \
OpSys                                                                 
Android             -            -           -       -            -   
Chrome OS           -       1559.0           -       -       2199.0   
Linux               -   811.638125  389.056364       -            -   
Mac OS X            -     1311.994           -       -       1222.0   
No OS               -   782.989286  404.675385       -       1098.5   
Windows 10     3100.0  1334.456872  664.717647       -  1911.985285   
Windows 10 S        -      1668.95           -       -       2589.0   
Windows 7           -    1846.4668  924.048333  1002.0  2235.396667   
macOS               -      1600.37           -       -      2180.87   

Memory_Amount         1024      2048  
OpSys                                 
Android                  -         -  
Chrome OS                -         -  
Linux            581.40129         -  
Mac OS X                 -         -  
No OS             540.5396     594.0  
Windows 10      904.827906  666.9475  
Windows 10 S             -         -  
Windows 7      1539.666667         -  
macOS                    -         -

# Задание 5
## 5.1 Ноутбуков каких компаний и с каким процессором больше?

In [44]:
stat.pivot_table(index = 'Company', columns = 'Cpu_Company',
                values = 'Price_euros', aggfunc = 'count',
                fill_value = '-')

Cpu_Company   AMD  Intel Samsung
Company                         
Acer         10.0   93.0       -
Apple           -   21.0       -
Asus         11.0  147.0       -
Chuwi           -    3.0       -
Dell            -  297.0       -
Fujitsu         -    3.0       -
Google          -    3.0       -
HP           25.0  249.0       -
Huawei          -    2.0       -
LG              -    3.0       -
Lenovo       16.0  281.0       -
MSI             -   54.0       -
Mediacom        -    7.0       -
Microsoft       -    6.0       -
Razer           -    7.0       -
Samsung         -    8.0     1.0
Toshiba         -   48.0       -
Vero            -    4.0       -
Xiaomi          -    4.0       -

In [45]:
pd.crosstab(index = stat['Company'], columns = stat["Cpu_Company"])

Cpu_Company  AMD  Intel  Samsung
Company                         
Acer          10     93        0
Apple          0     21        0
Asus          11    147        0
Chuwi          0      3        0
Dell           0    297        0
Fujitsu        0      3        0
Google         0      3        0
HP            25    249        0
Huawei         0      2        0
LG             0      3        0
Lenovo        16    281        0
MSI            0     54        0
Mediacom       0      7        0
Microsoft      0      6        0
Razer          0      7        0
Samsung        0      8        1
Toshiba        0     48        0
Vero           0      4        0
Xiaomi         0      4        0

## 5.2 С каким типом памяти и с каким объемом памяти больше ноутбуков?

In [46]:
stat.head(2)

Company      Product   TypeName  Inches                   Cpu  Ram  \
0   Apple  MacBook Pro  Ultrabook    13.3  Intel Core i5 2.3GHz  8GB   
1   Apple  Macbook Air  Ultrabook    13.3  Intel Core i5 1.8GHz  8GB   

                            Gpu  OpSys  Weight  Price_euros Cpu_Company  \
0  Intel Iris Plus Graphics 640  macOS    1.37      1339.69       Intel   
1        Intel HD Graphics 6000  macOS    1.34       898.94       Intel   

   Memory_Amount   Memory_Type  SSD  laptop_id  
0            128           SSD    1          0  
1            128  FlashStorage    0          1

In [47]:
pd.crosstab(index = stat['Memory_Type'], columns = stat['Memory_Amount'])

Memory_Amount    8     16    32    64    128   180   240   256   500   508   \
Memory_Type                                                                   
FlashStorage        0     7    38    15     4     0     0     8     0     0   
FlashStorage+1T     0     0     0     1     0     0     0     0     0     0   
HDD                 0     0     1     0     1     0     0     0   132     0   
HDD+1T              0     0     0     0     0     0     0     0     0     0   
Hybrid              0     0     0     0     0     0     0     0     0     1   
SSD                 1     3     6     1    76     5     1   412     0     0   
SSD+1.0T            0     0     0     0     0     0     0     1     0     0   
SSD+1T              0     0     0     0    94     0     0    73     0     0   
SSD+256G            0     0     0     0     0     0     0     2     0     0   
SSD+2T              0     0     0     0     2     0     0    10     0     0   
SSD+500G            0     0     0     0     0     0     0     2     0     0   
SSD+512G            0     0     0     0     0     0     0     0     0     0   

Memory_Amount    512   1024  2048  
Memory_Type                        
FlashStorage        2     0     0  
FlashStorage+1T     0     0     0  
HDD                 0   224    16  
HDD+1T              0     1     0  
Hybrid              0     9     0  
SSD               118    14     0  
SSD+1.0T            1     0     0  
SSD+1T             14     2     0  
SSD+256G            1     0     0  
SSD+2T              3     0     0  
SSD+500G            0     0     0  
SSD+512G            1     0     0

In [48]:
stat.pivot_table(index = 'Memory_Type', columns = 'Memory_Amount',
                values = 'Company', aggfunc = 'count',
                fill_value = '-')

Memory_Amount   8    16    32    64    128  180  240    256    500  508   \
Memory_Type                                                                
FlashStorage       -  7.0  38.0  15.0   4.0    -    -    8.0      -    -   
FlashStorage+1T    -    -     -   1.0     -    -    -      -      -    -   
HDD                -    -   1.0     -   1.0    -    -      -  132.0    -   
HDD+1T             -    -     -     -     -    -    -      -      -    -   
Hybrid             -    -     -     -     -    -    -      -      -  1.0   
SSD              1.0  3.0   6.0   1.0  76.0  5.0  1.0  412.0      -    -   
SSD+1.0T           -    -     -     -     -    -    -    1.0      -    -   
SSD+1T             -    -     -     -  94.0    -    -   73.0      -    -   
SSD+256G           -    -     -     -     -    -    -    2.0      -    -   
SSD+2T             -    -     -     -   2.0    -    -   10.0      -    -   
SSD+500G           -    -     -     -     -    -    -    2.0      -    -   
SSD+512G           -    -     -     -     -    -    -      -      -    -   

Memory_Amount     512    1024  2048  
Memory_Type                          
FlashStorage       2.0      -     -  
FlashStorage+1T      -      -     -  
HDD                  -  224.0  16.0  
HDD+1T               -    1.0     -  
Hybrid               -    9.0     -  
SSD              118.0   14.0     -  
SSD+1.0T           1.0      -     -  
SSD+1T            14.0    2.0     -  
SSD+256G           1.0      -     -  
SSD+2T             3.0      -     -  
SSD+500G             -      -     -  
SSD+512G           1.0      -     -

# Домашнее задание 1
## Скачать данные

In [49]:
import pandas as pd
import numpy as np
hd = pd.read_csv('/Users/macbookpro/Downloads/kc_house_data.csv', sep =',')

In [50]:
hd.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

## 1.1 Создать новый признак price_per_sq_lot, который будет содержать среднюю стоимость за один кв. метр общей площади

In [51]:
hd['price_per_sq_lot'] = hd['price']/hd['sqft_lot']

In [52]:
hd.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...        1180              0   
1      7242     2.0           0     0  ...        2170            400   
2     10000     1.0           0     0  ...         770              0   
3      5000     1.0           0     0  ...        1050            910   
4      8080     1.0           0     0  ...        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  price_per_sq_lot  
0        5650         39.274336  
1        7639         74.288870  
2        8062         18.000000  
3        5000        120.800000  
4        7503         63.118812  

[5 rows x 22 columns]

## 1.2 Создать новый признак delta_renovated, который будет содержать разницу в годах между годом реновацией дома и годом постройки дома

In [53]:
hd['delta_renovated'] = hd['yr_renovated'] - hd['yr_built']

In [54]:
hd.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  sqft_basement  yr_built  \
0      5650     1.0           0     0  ...              0      1955   
1      7242     2.0           0     0  ...            400      1951   
2     10000     1.0           0     0  ...              0      1933   
3      5000     1.0           0     0  ...            910      1965   
4      8080     1.0           0     0  ...              0      1987   

   yr_renovated  zipcode      lat     long  sqft_living15  sqft_lot15  \
0             0    98178  47.5112 -122.257           1340        5650   
1          1991    98125  47.7210 -122.319           1690        7639   
2             0    98028  47.7379 -122.233           2720        8062   
3             0    98136  47.5208 -122.393           1360        5000   
4             0    98074  47.6168 -122.045           1800        7503   

   price_per_sq_lot  delta_renovated  
0         39.274336            -1955  
1         74.288870               40  
2         18.000000            -1933  
3        120.800000            -1965  
4         63.118812            -1987  

[5 rows x 23 columns]

In [55]:
print(type(hd.iloc[1,22]))

<class 'numpy.int64'>


In [56]:
hd.loc[(hd.delta_renovated<0), 'delta_renovated'] = 0

In [57]:
hd.tail()

id             date     price  bedrooms  bathrooms  \
21608   263000018  20140521T000000  360000.0         3       2.50   
21609  6600060120  20150223T000000  400000.0         4       2.50   
21610  1523300141  20140623T000000  402101.0         2       0.75   
21611   291310100  20150116T000000  400000.0         3       2.50   
21612  1523300157  20141015T000000  325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  sqft_basement  \
21608         1530      1131     3.0           0     0  ...              0   
21609         2310      5813     2.0           0     0  ...              0   
21610         1020      1350     2.0           0     0  ...              0   
21611         1600      2388     2.0           0     0  ...              0   
21612         1020      1076     2.0           0     0  ...              0   

       yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
21608      2009             0    98103  47.6993 -122.346           1530   
21609      2014             0    98146  47.5107 -122.362           1830   
21610      2009             0    98144  47.5944 -122.299           1020   
21611      2004             0    98027  47.5345 -122.069           1410   
21612      2008             0    98144  47.5941 -122.299           1020   

       sqft_lot15  price_per_sq_lot  delta_renovated  
21608        1509        318.302387                0  
21609        7200         68.811285                0  
21610        2007        297.852593                0  
21611        1287        167.504188                0  
21612        1357        302.044610                0  

[5 rows x 23 columns]

## 1.3 Создайте признаки года продажи, месяца продажи

In [58]:
hd['yr_of_sale'] = 0
hd['month_of_sale'] = 0

In [96]:
hd.head(2)

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   

   sqft_lot  floors  waterfront  view  ...  yr_renovated  zipcode      lat  \
0      5650     1.0           0     0  ...             0    98178  47.5112   
1      7242     2.0           0     0  ...          1991    98125  47.7210   

      long  sqft_living15  sqft_lot15  price_per_sq_lot  delta_renovated  \
0 -122.257           1340        5650         39.274336                0   
1 -122.319           1690        7639         74.288870               40   

   yr_of_sale  month_of_sale  
0  0.20141013              0  
1  0.20141209              0  

[2 rows x 25 columns]

In [97]:
def sale_date(x):
    date = x.split('T')[0]
    return date

In [156]:
# def convert_to_gb(x):
#     memory = x.split(' ')[0]
#     if memory.endswith('GB'):
#         res = memory.replace('GB', '')
#     elif memory.endswith('TB'):
#         res = float(memory.replace ('TB', ''))*1024
#     return int(res)

In [214]:
hd['yr_of_sale'] = hd['date'].apply(sale_date)
hd['yr_of_sale'] = hd['yr_of_sale'].astype('int')

In [215]:
def year(x):
    res = x/10000
    return res

In [216]:
hd['yr_of_sale'] = hd['yr_of_sale'].apply(year)

In [217]:
hd['yr_of_sale'] = hd['yr_of_sale'].astype('str')

In [218]:
def sep(x):
    res = x.split('.')[0]
    return res

In [219]:
hd['yr_of_sale'] = hd['yr_of_sale'].apply(sep)

In [220]:
hd.tail()

id             date     price  bedrooms  bathrooms  \
21608   263000018  20140521T000000  360000.0         3       2.50   
21609  6600060120  20150223T000000  400000.0         4       2.50   
21610  1523300141  20140623T000000  402101.0         2       0.75   
21611   291310100  20150116T000000  400000.0         3       2.50   
21612  1523300157  20141015T000000  325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  yr_renovated  \
21608         1530      1131     3.0           0     0  ...             0   
21609         2310      5813     2.0           0     0  ...             0   
21610         1020      1350     2.0           0     0  ...             0   
21611         1600      2388     2.0           0     0  ...             0   
21612         1020      1076     2.0           0     0  ...             0   

       zipcode      lat     long  sqft_living15  sqft_lot15  price_per_sq_lot  \
21608    98103  47.6993 -122.346           1530        1509        318.302387   
21609    98146  47.5107 -122.362           1830        7200         68.811285   
21610    98144  47.5944 -122.299           1020        2007        297.852593   
21611    98027  47.5345 -122.069           1410        1287        167.504188   
21612    98144  47.5941 -122.299           1020        1357        302.044610   

       delta_renovated  yr_of_sale  month_of_sale  
21608                0        2014              5  
21609                0        2015              2  
21610                0        2014              6  
21611                0        2015              1  
21612                0        2014             10  

[5 rows x 25 columns]

In [221]:
hd['month_of_sale'] = hd['date'].apply(sale_date)
hd['month_of_sale'] = hd['month_of_sale'].astype('int')

In [222]:
def month(x):
    res = x%10000/100
    return res

In [223]:
hd['month_of_sale'] = hd['month_of_sale'].apply(month)
hd['month_of_sale'] = hd['month_of_sale'].astype('str')

In [224]:
hd['month_of_sale'] = hd['month_of_sale'].apply(sep)

In [225]:
hd.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  yr_renovated  zipcode      lat  \
0      5650     1.0           0     0  ...             0    98178  47.5112   
1      7242     2.0           0     0  ...          1991    98125  47.7210   
2     10000     1.0           0     0  ...             0    98028  47.7379   
3      5000     1.0           0     0  ...             0    98136  47.5208   
4      8080     1.0           0     0  ...             0    98074  47.6168   

      long  sqft_living15  sqft_lot15  price_per_sq_lot  delta_renovated  \
0 -122.257           1340        5650         39.274336                0   
1 -122.319           1690        7639         74.288870               40   
2 -122.233           2720        8062         18.000000                0   
3 -122.393           1360        5000        120.800000                0   
4 -122.045           1800        7503         63.118812                0   

   yr_of_sale  month_of_sale  
0        2014             10  
1        2014             12  
2        2015              2  
3        2014             12  
4        2015              2  

[5 rows x 25 columns]

## 1.4 Удалите признаки date, zipcode, lat, long

In [229]:
hd.drop(columns = ['date','zipcode','lat', 'long'], inplace = True)

In [230]:
hd.head()

id     price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0  7129300520  221900.0         3       1.00         1180      5650     1.0   
1  6414100192  538000.0         3       2.25         2570      7242     2.0   
2  5631500400  180000.0         2       1.00          770     10000     1.0   
3  2487200875  604000.0         4       3.00         1960      5000     1.0   
4  1954400510  510000.0         3       2.00         1680      8080     1.0   

   waterfront  view  condition  ...  sqft_above  sqft_basement  yr_built  \
0           0     0          3  ...        1180              0      1955   
1           0     0          3  ...        2170            400      1951   
2           0     0          3  ...         770              0      1933   
3           0     0          5  ...        1050            910      1965   
4           0     0          3  ...        1680              0      1987   

   yr_renovated  sqft_living15  sqft_lot15  price_per_sq_lot  delta_renovated  \
0             0           1340        5650         39.274336                0   
1          1991           1690        7639         74.288870               40   
2             0           2720        8062         18.000000                0   
3             0           1360        5000        120.800000                0   
4             0           1800        7503         63.118812                0   

   yr_of_sale month_of_sale  
0        2014            10  
1        2014            12  
2        2015             2  
3        2014            12  
4        2015             2  

[5 rows x 21 columns]

# Домашнее задание 2
## Создайте датафрейм с клиентами

In [262]:
clients_re = pd.DataFrame({
    'client_id': [1459, 4684, 3498, 3942, 4535, 2308, 2866, 2765, 1472, 4236, 2295, 939, 3840,  280,   20, 4332, 3475, 4213, 3113, 4809, 2134, 2242, 2068, 4929, 1384, 1589, 3317, 2260, 1727, 1764, 1611, 1474],
    'house_id': [8965450190, 6823100225, 5104540330, 2131701075, 1522700060, 1189000207, 6821600300, 7137950720, 9510920050, 6131600255, 5428000070, 1788800910, 8100400160, 3123049142, 6306800010, 5083000375, 7920100025, 1951600150,  809001400,  339600110, 1622049154, 1099600250, 8563000110, 2768100205, 3995700435, 8861700030, 3303980210, 7731100066, 8146100580,  825069097, 3889100029, 9524100196]
})

In [263]:
clients_re.head()

client_id    house_id
0       1459  8965450190
1       4684  6823100225
2       3498  5104540330
3       3942  2131701075
4       4535  1522700060

## 2.1 Присоедините к таблице clients данные по домам через метод join

In [264]:
house_id = clients_re.set_index('house_id')

In [265]:
hd_id = hd.set_index('id')

In [287]:
joined = house_id.join(hd_id)
joined

client_id      price  bedrooms  bathrooms  sqft_living  sqft_lot  \
339600110        4809   395000.0         3       2.50         1610      3755   
809001400        3113   925000.0         3       1.00         1630      3200   
825069097        1764   770000.0         3       2.50         2650     40705   
1099600250       2242   260000.0         3       1.75         1710      6400   
1189000207       2308   387000.0         2       2.50         1170      1394   
1522700060       4535   518000.0         4       2.75         2520     14021   
1622049154       2134   289900.0         3       1.75         1899     11325   
1788800910        939   190000.0         3       1.00         1200     10458   
1951600150       4213   180000.0         3       1.00         1610      8500   
2131701075       3942   420000.0         3       1.75         1720      5000   
2768100205       4929   519000.0         4       2.50         1950      2617   
3123049142        280   452000.0         3       2.25         2600     14810   
3303980210       3317  1115000.0         4       3.75         4040     14212   
3889100029       1611   810000.0         3       2.50         2670     10481   
3995700435       1384   265000.0         4       3.00         1940      8170   
5083000375       4332   170000.0         3       1.00         1310      9529   
5083000375       4332   235000.0         3       1.00         1310      9529   
5104540330       3498   679000.0         4       2.50         3680      7236   
5428000070       2295   770000.0         5       3.50         4750      8234   
6131600255       4236   202500.0         3       2.00         1540      8316   
6306800010         20   436472.0         4       2.50         2692      8392   
6821600300       2866   886000.0         3       2.25         2380      6000   
6823100225       4684   700000.0         4       1.75         1870      6000   
7137950720       2765   339100.0         4       2.50         2350     10655   
7731100066       2260   545000.0         3       1.00         1510      5000   
7920100025       3475   450000.0         2       1.00          740      5100   
8100400160       3840   700000.0         3       2.25         2330     11424   
8146100580       1727   765000.0         3       1.00         1220      7585   
8563000110       2068   427000.0         4       1.75         1460      9750   
8861700030       1589   510000.0         3       1.50         2400     10275   
8965450190       1459   295000.0         3       2.50         1500      3060   
9510920050       1472   725000.0         3       2.50         2980     16996   
9524100196       1474   239000.0         2       1.50          680       772   

            floors  waterfront  view  condition  ...  sqft_above  \
339600110      2.0           0     0          3  ...        1610   
809001400      1.5           0     0          4  ...        1630   
825069097      2.0           0     0          3  ...        2650   
1099600250     1.0           0     0          4  ...        1240   
1189000207     2.0           0     0          3  ...        1170   
1522700060     2.0           0     0          3  ...        2520   
1622049154     2.0           0     0          3  ...        1899   
1788800910     1.0           0     0          4  ...        1200   
1951600150     1.5           0     0          3  ...        1610   
2131701075     1.5           0     0          3  ...        1720   
2768100205     1.5           0     0          4  ...        1250   
3123049142     1.0           0     2          4  ...        1490   
3303980210     2.0           0     0          3  ...        4040   
3889100029     2.0           0     0          3  ...        2670   
3995700435     1.0           0     0          4  ...        1940   
5083000375     1.0           0     0          3  ...        1310   
5083000375     1.0           0     0          3  ...        1310   
5104540330     2.0           0     0          3  ...        3680   

## 2.2 Присоедините к таблице clients данные по домам через метод merge

In [289]:
merged = clients_re.merge(hd, left_on = 'house_id', right_on = 'id')
merged.drop(columns = ['id'], inplace = True)
merged.head()

client_id    house_id     price  bedrooms  bathrooms  sqft_living  \
0       1459  8965450190  295000.0         3       2.50         1500   
1       4684  6823100225  700000.0         4       1.75         1870   
2       3498  5104540330  679000.0         4       2.50         3680   
3       3942  2131701075  420000.0         3       1.75         1720   
4       4535  1522700060  518000.0         4       2.75         2520   

   sqft_lot  floors  waterfront  view  ...  sqft_above  sqft_basement  \
0      3060     2.0           0     0  ...        1500              0   
1      6000     1.0           0     0  ...        1670            200   
2      7236     2.0           0     0  ...        3680              0   
3      5000     1.5           0     0  ...        1720              0   
4     14021     2.0           0     0  ...        2520              0   

   yr_built  yr_renovated  sqft_living15  sqft_lot15  price_per_sq_lot  \
0      1994             0           2700        7734         96.405229   
1      1949             0           1710        6000        116.666667   
2      2006             0           3310        7180         93.836374   
3      1932          2009           1410        8300         84.000000   
4      1999             0           2330       14007         36.944583   

   delta_renovated  yr_of_sale  month_of_sale  
0                0        2015              2  
1                0        2015              4  
2                0        2015              5  
3               77        2014             12  
4                0        2014              6  

[5 rows x 22 columns]

# Домашнее задание 3
## 3.1 Найдите среднюю стоимость домов в зависимости от количества спален

In [290]:
hd.head()

id     price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0  7129300520  221900.0         3       1.00         1180      5650     1.0   
1  6414100192  538000.0         3       2.25         2570      7242     2.0   
2  5631500400  180000.0         2       1.00          770     10000     1.0   
3  2487200875  604000.0         4       3.00         1960      5000     1.0   
4  1954400510  510000.0         3       2.00         1680      8080     1.0   

   waterfront  view  condition  ...  sqft_above  sqft_basement  yr_built  \
0           0     0          3  ...        1180              0      1955   
1           0     0          3  ...        2170            400      1951   
2           0     0          3  ...         770              0      1933   
3           0     0          5  ...        1050            910      1965   
4           0     0          3  ...        1680              0      1987   

   yr_renovated  sqft_living15  sqft_lot15  price_per_sq_lot  delta_renovated  \
0             0           1340        5650         39.274336                0   
1          1991           1690        7639         74.288870               40   
2             0           2720        8062         18.000000                0   
3             0           1360        5000        120.800000                0   
4             0           1800        7503         63.118812                0   

   yr_of_sale month_of_sale  
0        2014            10  
1        2014            12  
2        2015             2  
3        2014            12  
4        2015             2  

[5 rows x 21 columns]

In [304]:
result = hd.groupby('bedrooms').agg({'price': 'mean'})
result['price'] = result['price'].astype(int)

In [305]:
result.sort_values('price', ascending=True)

price
bedrooms         
1          317642
2          401372
0          409503
3          466232
11         520000
4          635419
33         640000
5          786599
10         819333
6          825520
9          893999
7          951184
8         1105076

## 3.2 Найдите минимальную, среднюю и максимальную стоимости домов в зависимости от состояния дома

In [309]:
result_condition = hd.groupby('condition').agg({'price': ['min', 'mean', 'max']})
result_condition['price'] = result_condition['price'].astype(int)
result_condition

price                 
              min    mean      max
condition                         
1           78000  334431  1500000
2           80000  327287  2555000
3           75000  542012  7062500
4           89000  521200  7700000
5          110000  612418  3650000

## 3.3 Постройте таблицу с подсчетом количества домов в данных в зависимости от вида на набережную и оценкой вида

In [312]:
hd.head(2)

id     price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0  7129300520  221900.0         3       1.00         1180      5650     1.0   
1  6414100192  538000.0         3       2.25         2570      7242     2.0   

   waterfront  view  condition  ...  sqft_above  sqft_basement  yr_built  \
0           0     0          3  ...        1180              0      1955   
1           0     0          3  ...        2170            400      1951   

   yr_renovated  sqft_living15  sqft_lot15  price_per_sq_lot  delta_renovated  \
0             0           1340        5650         39.274336                0   
1          1991           1690        7639         74.288870               40   

   yr_of_sale month_of_sale  
0        2014            10  
1        2014            12  

[2 rows x 21 columns]

In [313]:
hd.pivot_table(index = 'waterfront', columns = 'view',
               values = 'id', aggfunc = 'count',
               fill_value='-'
)

view              0      1      2      3      4
waterfront                                     
0           19489.0  331.0  955.0  491.0  184.0
1                 -    1.0    8.0   19.0  135.0

## 3.4 Каких домов в зависимости от этажности и количества спален больше?

In [324]:
pd.crosstab(index = hd['bathrooms'], columns= hd['floors'])

floors      1.0  1.5   2.0  2.5  3.0  3.5
bathrooms                                
0.00          5    0     3    0    1    1
0.50          3    0     1    0    0    0
0.75         64    5     3    0    0    0
1.00       3115  624   106    3    4    0
1.25          2    2     3    0    2    0
1.50        904  205   268    5   64    0
1.75       2494  282   244    8   20    0
2.00       1283  282   304   12   49    0
2.25        883   95   940   15  114    0
2.50       1013  155  3966   39  204    3
2.75        500  120   542    7   15    1
3.00        228   76   391   23   33    2
3.25         72   25   430   12   50    0
3.50         58   17   614   13   29    0
3.75         18   11   113    7    6    0
4.00         18    4    99    5    9    1
4.25          3    2    67    4    3    0
4.50          5    3    81    3    8    0
4.75          3    2    14    3    1    0
5.00          4    0    16    1    0    0
5.25          1    0    12    0    0    0
5.50          3    0     7    0    0    0
5.75          0    0     4    0    0    0
6.00          0    0     6    0    0    0
6.25          0    0     2    0    0    0
6.50          0    0     2    0    0    0
6.75          1    0     1    0    0    0
7.50          0    0     1    0    0    0
7.75          0    0     1    0    0    0
8.00          0    0     0    1    1    0

In [317]:
hd.pivot_table(index = 'bathrooms', columns = 'floors',
               values = 'id', aggfunc = 'count',
               fill_value= '-'
)

floors        1.0    1.5     2.0   2.5    3.0  3.5
bathrooms                                         
0.00          5.0      -     3.0     -    1.0  1.0
0.50          3.0      -     1.0     -      -    -
0.75         64.0    5.0     3.0     -      -    -
1.00       3115.0  624.0   106.0   3.0    4.0    -
1.25          2.0    2.0     3.0     -    2.0    -
1.50        904.0  205.0   268.0   5.0   64.0    -
1.75       2494.0  282.0   244.0   8.0   20.0    -
2.00       1283.0  282.0   304.0  12.0   49.0    -
2.25        883.0   95.0   940.0  15.0  114.0    -
2.50       1013.0  155.0  3966.0  39.0  204.0  3.0
2.75        500.0  120.0   542.0   7.0   15.0  1.0
3.00        228.0   76.0   391.0  23.0   33.0  2.0
3.25         72.0   25.0   430.0  12.0   50.0    -
3.50         58.0   17.0   614.0  13.0   29.0    -
3.75         18.0   11.0   113.0   7.0    6.0    -
4.00         18.0    4.0    99.0   5.0    9.0  1.0
4.25          3.0    2.0    67.0   4.0    3.0    -
4.50          5.0    3.0    81.0   3.0    8.0    -
4.75          3.0    2.0    14.0   3.0    1.0    -
5.00          4.0      -    16.0   1.0      -    -
5.25          1.0      -    12.0     -      -    -
5.50          3.0      -     7.0     -      -    -
5.75            -      -     4.0     -      -    -
6.00            -      -     6.0     -      -    -
6.25            -      -     2.0     -      -    -
6.50            -      -     2.0     -      -    -
6.75          1.0      -     1.0     -      -    -
7.50            -      -     1.0     -      -    -
7.75            -      -     1.0     -      -    -
8.00            -      -       -   1.0    1.0    -

#### Больше всего домов на 2 этаже с 2.5 спальнями (3966 ед.)

## 3.5 Постройте таблицу с подсчетом медианной стоимости домов в данных в зависимости от состояния дома и оценки дома

In [318]:
hd.head(2)

id     price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0  7129300520  221900.0         3       1.00         1180      5650     1.0   
1  6414100192  538000.0         3       2.25         2570      7242     2.0   

   waterfront  view  condition  ...  sqft_above  sqft_basement  yr_built  \
0           0     0          3  ...        1180              0      1955   
1           0     0          3  ...        2170            400      1951   

   yr_renovated  sqft_living15  sqft_lot15  price_per_sq_lot  delta_renovated  \
0             0           1340        5650         39.274336                0   
1          1991           1690        7639         74.288870               40   

   yr_of_sale month_of_sale  
0        2014            10  
1        2014            12  

[2 rows x 21 columns]

In [321]:
hd.pivot_table(index = 'grade', columns = 'condition',
              values = 'price', aggfunc = 'median',
               fill_value = '-'
              )

condition         1          2          3          4          5
grade                                                          
1          142000.0          -          -          -          -
3                 -   280000.0    75000.0          -   262000.0
4          150000.0   145000.0   205000.0   238525.0          -
5          190000.0   180000.0   234475.0   229975.0   227450.0
6          255000.0   235000.0   265000.0   295000.0   285475.0
7          403500.0   305000.0   357500.0   390000.0   456000.0
8          932500.0   429000.0   485000.0   571250.0   696000.0
9                 -   715000.0   689000.0   823500.0  1078000.0
10                -  1752500.0   890000.0  1030000.0  1650000.0
11                -          -  1209500.0  1685000.0  2050000.0
12                -          -  1807500.0  2125000.0  1990000.0
13                -          -  2888000.0  5750000.0          -

In [323]:
pd.crosstab(index = hd['grade'], columns = hd['condition'])

condition   1   2     3     4    5
grade                             
1           1   0     0     0    0
3           0   1     1     0    1
4           1   5    13    10    0
5           9  15   100    84   34
6          11  59  1035   685  248
7           6  75  5234  2833  833
8           2  13  4269  1394  390
9           0   2  2041   446  126
10          0   2   921   156   55
11          0   0   332    56   11
12          0   0    74    13    3
13          0   0    11     2    0